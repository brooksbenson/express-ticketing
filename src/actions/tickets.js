import db from '../firebase/firebase';

export const addTicket = ({ key, ...ticket }) => ({
  type: 'ADD_TICKET',
  key,
  ticket
});

export const startAddTicket = ({ title, urgency }) => {
  return async (dispatch, getState) => {
    const state = getState();
    const ticket = {
      date: Date.now(),
      title,
      urgency,
      accountKey: state.activeAccountKey,
      contactKey: state.activeContactKey,
      userKeys: { [state.activeUserKey]: true }
    };
    const { key } = await db.ref('tickets').push(ticket);
    await Promise.all([
      db.ref('open_tickets').update({ [key]: true }),
      db.ref(`user_tickets/${state.activeUserKey}`).update({ [key]: true })
    ]);
    dispatch(addTicket({ key, ...ticket }));
    return key;
  };
};

export const setTickets = tickets => ({
  type: 'SET_TICKETS',
  tickets
});

export const startSetTickets = () => {
  return async (dispatch, getState) => {
    const state = getState();
    const [openTickets, userTickets] = await Promise.all([
      db.ref('open_tickets').once('value'),
      db.ref(`user_tickets/${state.activeUserKey}`).once('value')
    ]);
    const toFetch = [];
    const userTicketsObj = userTickets.val();
    Object.keys(openTickets.val()).forEach(key => {
      if (key in userTicketsObj) toFetch.push(key);
    });
    const tickets = {};
    await Promise.all(
      toFetch.map(
        key =>
          new Promise(resolve => {
            db.ref(`tickets/${key}`)
              .once('value')
              .then(snap => {
                tickets[key] = snap.val();
                resolve();
              });
          })
      )
    );
    dispatch(setTickets(tickets));
  };
};
