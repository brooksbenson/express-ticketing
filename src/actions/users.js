import db, { auth } from '../firebase/firebase';

export const addUser = ({ key, ...user }) => ({
  type: 'ADD_USER',
  key,
  user
});

export const startAddUser = user => {
  return async dispatch => {
    const { key } = await db.ref('user_data').push(user);
    dispatch(addUser({ key, ...user }));
    return key;
  };
};

export const setUsers = users => ({
  type: 'SET_USERS',
  users
});

export const startSetUsers = () => {
  return async dispatch => {
    const snap = await db.ref('user_data').once('value');
    dispatch(setUsers(snap.val()));
  };
};
