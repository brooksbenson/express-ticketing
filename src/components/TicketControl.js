import React from 'react';

export default props => (
  <div className={props.className}>
    <h3> Ticket Control </h3>
    <div>
      <h4> Triage </h4>
      <span> Urgency </span>
      <select
        className="select"
        onChange={e => props.onUrgencyChange(e.target.value)}
        value={props.urgency}
      >
        <option value="low">Low</option>
        <option value="medium">Medium</option>
        <option value="high">High</option>
      </select>
      <button onClick={props.onStatusChange}>
        {props.isOpen ? 'Close Ticket' : 'Reopen Ticket'}
      </button>
    </div>
    <div>
      <h4> Users </h4>
      <button onClick={props.openAddUserModal}> Add </button>
      <ul>{props.users.map(u => <li key={u.key}>{u.name}</li>)}</ul>
    </div>
  </div>
);
