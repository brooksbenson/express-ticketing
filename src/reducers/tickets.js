export default (state = [], action) => {
  switch (action.type) {
    case 'ADD_TICKET':
      return { ...state, [action.key]: action.ticket };
    case 'SET_TICKETS':
      return action.tickets;
    case 'UPDATE_URGENCY': {
      const { key, urgency } = action;
      const ticket = state[key];
      return { ...state, [key]: { ...ticket, urgency } };
    }
    case 'ADD_USER': {
      const { key, userKey } = action;
      const ticket = state[key];
      const userKeys = { ...ticket.userKeys, [userKey]: true };
      return { ...state, [key]: { ...ticket, userKeys } };
    }
    default:
      return state;
  }
};
