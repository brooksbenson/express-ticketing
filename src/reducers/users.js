export default (state = {}, action) => {
  switch (action.type) {
    case 'ADD_USER':
      return { ...state, [action.key]: action.user };
    case 'SET_USERS':
      return action.users;
    default:
      return state;
  }
};
