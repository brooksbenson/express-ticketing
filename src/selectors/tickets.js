export default (tickets, query) => {
  const regex = new RegExp(query.split('').join('.*?'), 'i');
  return tickets.filter(
    t =>
      regex.test(t.accountName) || regex.test(t.urgency) || regex.test(t.title)
  );
};
