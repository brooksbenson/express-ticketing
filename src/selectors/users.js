export default (users, search) => {
  const regex = new RegExp(search.split('').join('.*?'), 'i');
  return users.filter(user => regex.test(user.email));
};
