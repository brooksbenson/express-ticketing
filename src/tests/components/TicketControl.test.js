import React from 'react';
import { shallow } from 'enzyme';
import { ticketsArr } from '../fixtures/tickets';
import { usersObj } from '../fixtures/users';
import TicketControl from '../../components/TicketControl';

const { userKeys, urgency } = ticketsArr[2];

const users = userKeys.map(key => ({ ...usersObj[key], key }));
const className = 'ticket-control';
const isOpen = true;

let onUrgencyChange;
let onStatusChange;
let openAddUserModal;
let wrapper;

beforeEach(() => {
  onUrgencyChange = jest.fn();
  onStatusChange = jest.fn();
  openAddUserModal = jest.fn();
  wrapper = shallow(
    <TicketControl
      className={className}
      isOpen={isOpen}
      users={users}
      onUrgencyChange={onUrgencyChange}
      onStatusChange={onStatusChange}
      openAddUserModal={openAddUserModal}
    />
  );
});

test('TicketControl should match snapshot', () => {
  expect(wrapper).toMatchSnapshot();
});

test('TicketControl should give wrapper element the className passed as a prop', () => {
  expect(wrapper.hasClass(className)).toBeTruthy();
});

test('TicketControl should invoke onUrgencyChange on urgency change', () => {
  const value = 'high';
  wrapper.find('select').simulate('change', { target: { value } });
  expect(onUrgencyChange).toHaveBeenCalledWith(value);
});

test('TicketControl should render a list of users correctly', () => {
  const userListItems = wrapper.find('li');
  expect(userListItems.at(0).text()).toBe(users[0].name);
  expect(userListItems.at(1).text()).toBe(users[1].name);
});

test('TicketControl should invoke onStatusChange on close ticket button click', () => {
  wrapper
    .find('button')
    .at(0)
    .simulate('click');
  expect(onStatusChange).toHaveBeenCalled();
});

test('TicketControl should invoke openAddUserModal on add user button click', () => {
  wrapper
    .find('button')
    .at(1)
    .simulate('click');
  expect(openAddUserModal).toHaveBeenCalled();
});

test('TicketControl should render ticket status button with text "Close Ticket" when isOpen prop is true', () => {
  expect(
    wrapper
      .find('button')
      .at(0)
      .text()
  ).toBe('Close Ticket');
});

test('TicketControl should render ticket status button with text "Reopen Ticket" when isOpen prop is false', () => {
  wrapper.setProps({ isOpen: false });
  expect(
    wrapper
      .find('button')
      .at(0)
      .text()
  ).toBe('Reopen Ticket');
});
