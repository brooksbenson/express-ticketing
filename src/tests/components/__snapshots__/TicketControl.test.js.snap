// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TicketControl should match snapshot 1`] = `
<div
  className="ticket-control"
>
  <h3>
     Ticket Control 
  </h3>
  <div>
    <h4>
       Triage 
    </h4>
    <span>
       Urgency 
    </span>
    <select
      className="select"
      onChange={[Function]}
    >
      <option
        value="low"
      >
        Low
      </option>
      <option
        value="medium"
      >
        Medium
      </option>
      <option
        value="high"
      >
        High
      </option>
    </select>
    <button
      onClick={[MockFunction]}
    >
      Close Ticket
    </button>
  </div>
  <div>
    <h4>
       Users 
    </h4>
    <button
      onClick={[MockFunction]}
    >
       Add 
    </button>
    <ul>
      <li
        key="-keiajf"
      >
        Brooks Benson
      </li>
      <li
        key="-wqqadv"
      >
        Jeff Benson
      </li>
    </ul>
  </div>
</div>
`;
